feat: Add kind as a service

adrianriobo opened the pull request:
Fix #462 

This is based on #455 to reuse some refactoring

Dannyb48 commented on pkg/util/cloud-config/kind/cloud-config:
For Konflux, since it is a pretty beefy platform and depending on the pipelines runs we suggest to somtimes bump memory reserves to 12g[1]. Any chance we can make this configurable? either somehow pass our own kind-config file on the cli as an arg or specific params that modifiy certain attributes on this default kind config file on the cli ?

[1]https://github.com/konflux-ci/konflux-ci/tree/c0d8afb12ab8fb7677bcf90337e1630dc69f48b8?tab=readme-ov-file#bootstrapping-the-cluster

adrianriobo replied:
sure that one is easy

flacatus commented:
after testing few times i got the following error:

``` log
All requirements are met
Continue
Testing PVC creation for default storage class
Creating PVC from './dependencies/pre-deployment-pvc-binding' using the cluster's default storage class
error: error validating "./dependencies/pre-deployment-pvc-binding": error validating data: failed to download openapi: Get "https://127.0.0.1:33397/openapi/v2?timeout=32s": dial tcp 127.0.0.1:33397: connect: connection refused; if you choose to ignore these errors, turn validation off with --validate=false
error: error validating "./dependencies/pre-deployment-pvc-binding": error validating data: failed to download openapi: Get "https://127.0.0.1:33397/openapi/v2?timeout=32s": dial tcp 127.0.0.1:33397: connect: connection refused; if you choose to ignore these errors, turn validation off with --validate=false
error: error validating "./dependencies/pre-deployment-pvc-binding": error validating data: failed to download openapi: Get "https://127.0.0.1:33397/openapi/v2?timeout=32s": dial tcp 127.0.0.1:33397: connect: connection refused; if you choose to ignore these errors, turn validation off with --validate=false
kubectl apply -k ./dependencies/pre-deployment-pvc-binding: Error while creating pre-deployment-pvc-binding.
```

flacatus commented:
```log
[36mINFO[0m  +  aws:autoscaling:Group main-akd-asg creating (9s) error:   sdk-v2/provider2.go:509: sdk.helper_schema: waiting for Auto Scaling Group (main-akd-asg-663e6e0) capacity satisfied: scaling activity (860658d9-07ab-be97-a7ff-b87e1f0854bf): Failed: Could not launch Spot Instances. UnfulfillableCapacity - Unable to fulfill capacity due to your request configuration. Please adjust your request and try again. Launching EC2 instance failed.: provider=aws@6.75.0 
[36mINFO[0m  +  aws:autoscaling:Group main-akd-asg creating (9s) error: 1 error occurred: 
[36mINFO[0m  +  aws:autoscaling:Group main-akd-asg **creating failed** error: 1 error occurred: 
[36mINFO[0m @ updating.............                      
[36mINFO[0m  +  command:remote:Command main-ocp-readiness-akd-cmd creating (30s) Dial 2/inf failed: retrying 
[36mINFO[0m @ updating..................                 
[36mINFO[0m  +  command:remote:Command main-ocp-readiness-akd-cmd creating (46s) Dial 3/inf failed: retrying 
[36mINFO[0m @ updating...................                
[36mINFO[0m  +  command:remote:Command main-ocp-readiness-akd-cmd creating (61s) Dial 4/inf failed: retrying 
[36mINFO[0m @ updating..................                 
[36mINFO[0m  +  command:remote:Command main-ocp-readiness-akd-cmd creating (76s) Dial 5/inf failed: retrying 
[36mINFO[0m @ updating...................                
[36mINFO[0m  +  command:remote:Command main-ocp-readiness-akd-cmd creating (92s) Dial 6/inf failed: retrying 
[36mINFO[0m @ updating................                   
[36mINFO[0m  +  aws:ec2:NatGateway natgateway-publicmain-akd-net0 created (106s)  
[36mINFO[0m @ updating.....                              
[36mINFO[0m  +  command:remote:Command main-ocp-readiness-akd-cmd creating (107s) Dial 7/inf failed: retrying 
```

if failed to launch the machine is continuing to do operations

adrianriobo commented:
The Kubeconfig issue is due to some random issue on kind generating it, but still adjusting the string to be replaced with the actual value of Public IP and fixed 6643 port works, as so I changed the replace string on kubeconfig.

For the provisioning issue we added a safe threshold to be added to spot price, previously we had it at 10%, increasing it to 20% I see no failure even on bigger instances. In the future this parameter will be customize through parameter

flacatus commented:
I see main-ocp-readiness to main-kubernetes-readiness

flacatus commented:
```
[36mINFO[0m  +  aws:autoscaling:Group main-akd-asg creating (24s) error:   sdk-v2/provider2.go:509: sdk.helper_schema: waiting for Auto Scaling Group (main-akd-asg-abb6eec) capacity satisfied: scaling activity (70465abd-e486-9946-6141-aba9a1d0f220): Failed: Could not launch Spot Instances. UnfulfillableCapacity - Unable to fulfill capacity due to your request configuration. Please adjust your request and try again. Launching EC2 instance failed.: provider=aws@6.77.0
[36mINFO[0m  +  aws:autoscaling:Group main-akd-asg creating (24s) error: 1 error occurred:
[36mINFO[0m  +  aws:autoscaling:Group main-akd-asg **creating failed** error: 1 error occurred:
```

I still see issues when using big instances like:
```yaml
          - name: cpus
            value: '48'
          - name: memory
            value: '192'
```

flacatus commented:
threshold maybe is better to expose this param to tekton tasks...
